src/tracer.js

```diff
@@ -253,6 +270,32 @@ export default class Tracer {
     }
 
     /**
+     * Reapplies sampling if the parent span isn't sampled
+     *
+     * @param parent the parent span context
+     * @param ctx the span context
+     * @param operationName the operation name for the child
+     * @param tags tags to be applied by the sampler
+     */
+    handleUpsampling(parent: SpanContext, ctx: SpanContext, operationName: string, tags: any) {
+        if (!this._upsampling.enabled) {
+            return
+        }
+
+        // Reuse upsampling decision, if one exists in parent
+        if (parent.upsamplingDecision) {
+            ctx.samplingDecision = parent.upsamplingDecision.valueOf();
+            ctx.upsamplingDecision = parent.upsamplingDecision;
+            return
+        }
+
+        // Store upsampling decision in parent
+        parent.upsamplingDecision = new Boolean(this._sampler.isSampled(operationName, tags));
+        ctx.samplingDecision = parent.upsamplingDecision.valueOf();
+        ctx.upsamplingDecision = parent.upsamplingDecision;
+    }
```

Comment:
`debug` flag makes all other flags irrelevant. It implies finalized=true.

`sampled` flag does not imply finalized, but if sampled=true, it can never become false, and it will eventually become finalized.

Upsampling can only happen
  * on the *first span in process*
  * which is *not* a root span (otherwise regular sampling happens)
  * if that span has `sampled=false`

But it can happen at most twice:
  * in `startSpan` where currently we simply inherit parent `flags` and finalize both parent and new span. Instead, we can make a sampling decision here (subject to conditions above). If that decision is Yes then upsampling use case is done, it will never come up again. If that decision is No (i.e. we didn't change sampled=false inherited from parent), then it can be retried in setOperationName. The other difference here is that previously we were always finalizing the child span. Now we can skip finalizing if we match the upsampling criteria.
  * in setOperationName the difference now is that we could reach it for a child span that's not finalized (was impossible before). We'd still have sampled=false as otherwise it doesn't matter. I don't think we need any changes there, it would "just work" and potentially upsample the span due to it not being finalized (but it would only apply to first-in-process span per above conditions.